### DISCUSSION TITLE
BUG CHANGE ID FOR WINDOWS CLONE

### DISCUSSION THREAD

**Discussion by dokomade1111** [rustdesk/rustdesk#5648] [General]
### Bug Description

I have a Windows OS which has Rustdesk installed. but after I installed it on several computers then I checked if there were DUPLICATE IDs / ALL THE SAME IDs. finally I couldn't use rustdesk

### How to Reproduce

I don't have a solution. help me to change ID. I see that TEAM VIEWER is easier to change ID. If Rustdesk is an alternative to TEAM VIEWER, changing the ID should be very easy to do

### Expected Behavior

please fix problem

### Operating system(s) on local side and remote side

WINDOWS 10

### RustDesk Version(s) on local side and remote side

new version

### Screenshots

HAS THE SAME ID AS WINDOWS CLONE

### Additional Context

please fix problem

**Comment by dinger1986** [comment]
you of course are regenerating sids etc once you have dont your clones so its essentially a different machine? Like doing clones properly?

**Comment by dokomade1111** [comment]
@dinger1986 do you have to regenerate sids after cloning so that the RUSTDESK ID changes?

**Comment by dinger1986** [comment]
maybe, it needs to know its a different machine anyway

**Comment by dokomade1111** [comment]
@dinger1986 the solution was very difficult. because changing the sids I see is difficult too

**Comment by ghost** [comment]
I had the exactly same issue. Upon cloning a Windows virtual machine, they both ended with the same ID. So far, any attempts to change the ID were not successful. When attempting to connect to that ID, it (seems) randomly connects to one of the machines. The only work around that worked at this situation was to start using the local IP of the certain machine.

**Reply by abdullah-erturk** [reply]
https://github.com/abdullah-erturk/RustDesk-ID-Changer

**Reply by ghost** [reply]
Excellent work, thank you!

Tried the 3 scripts on Windows 2012 and they work. The first 2 worked as expected. The first used the host name as ID and second generated a random 9 digit ID. The third one, I entered 5 letters as the new ID and it ended with a random ID. I assume, the minimum is 9 digits/letters and if they are not there, it creates random. If so, this is a good workaround if the user is not aware for the length of the ID.

Great work, thank you.

**Reply by Zixim** [reply]
I've set a six character ID on multiple hosts.

**Comment by SergeyMy** [comment]
strange, how did you get the same identifier? essentially this is the **MAC address** of the network card :)
![изображение](https://github.com/rustdesk/rustdesk/assets/131688106/d5972200-63ed-42cb-a177-5ffb55033bd1)

**Comment by Zixim** [comment]
I'm having the same issue, on Rustdesk 1.2.0
Some way to manually change the ID would be most welcome.

**Comment by abdullah-erturk** [comment]
https://github.com/abdullah-erturk/RustDesk-ID-Changer

**Comment by abdullah-erturk** [comment]
Everything is automatic and all options are in one place

https://github.com/abdullah-erturk/RustDesk-ID-Changer

![sample](https://raw.githubusercontent.com/abdullah-erturk/RustDesk-ID-Changer/main/RustDesk_ID_Changer.jpg)

**Comment by Zixim** [comment]
@abdullah-erturk Why are you obfuscating code in that script ?

**Reply by abdullah-erturk** [reply]
I didn't encrypt the code, I just embedded it in the bat file with the project I linked below.

https://github.com/AveYo/Compressed2TXT

Thus, all files became one file.

If you want to be sure, you can look in the %temp% folder. All ps1 files are extracted there and run from there.

**Reply by Zixim** [reply]
Yes you do encrypt the code, why else would you be using a decryption key ?
```
$key = '*)eHY0=5tL>1K.8uphov/db;^+qWZQ2&N`cwgAkOnr~]Sf$Elma6U#RXzJ9IV4,-C?%x@!(}D\j7y[{GsT3iP_|MB<F'
```
<details>
  <summary>click here to see the deobfuscated code</summary>
  
  This is what ChatGPT returns when prompted to "make the source code human-readable"...
```
# Define a key for decryption
$key = '*)eHY0=5tL>1K.8uphov/db;^+qWZQ2&N`cwgAkOnr~]Sf$Elma6U#RXzJ9IV4,-C?%x@!(}D\j7y[{GsT3iP_|MB<F'

# Add a C# class to decrypt a file
Add-Type -TypeDefinition @'
using System.IO;

public class Decryptor
{
    public static void Decrypt(ref string[] files, int index, string outputFile, string key)
    {
        unchecked
        {
            int n = 0, c = 255, q = 0, v = 91;
            int z = files[index].Length;
            byte[] b91 = new byte[256];

            while (c > 0)
                b91[c--] = 91;

            while (c < 91)
                b91[key[c]] = (byte)c++;

            using (FileStream output = new FileStream(outputFile, FileMode.Create))
            {
                for (int i = 0; i != z; i++)
                {
                    c = b91[files[index][i]];

                    if (c == 91)
                        continue;

                    if (v == 91)
                    {
                        v = c;
                    }
                    else
                    {
                        v += c * 91;
                        q |= v << n;

                        if ((v & 8191) > 88)
                        {
                            n += 13;
                        }
                        else
                        {
                            n += 14;
                        }

                        v = 91;
                        do
                        {
                            output.WriteByte((byte)q);
                            q >>= 8;
                            n -= 8;
                        } while (n > 7);
                    }
                }

                if (v != 91)
                    output.WriteByte((byte)(q | v << n));
            }
        }
    }
}
'@

# Change the working directory
cd -LiteralPath ($env:__CD__)

# Define a function to decrypt a file
function DecryptFile([int]$fileIndex = 1)
{
    [Decryptor]::Decrypt([ref]$files, $fileIndex + 1, $fileIndex, $key)
    Expand-Archive -Path $fileIndex -Force -DestinationPath .  # Extract the decrypted file
    Remove-Item $fileIndex -Force  # Delete the encrypted file
}
```
</details>

Even though your intentions aren't malicious, it really isn't a good idea to spread code that can't easily be inspected.

All are free to make their own conclusions. I will not run obfuscated/encrypted powershell code on any of my hosts.

**Reply by abdullah-erturk** [reply]
This is not encryption, it is the process of embedding multiple files.

Here's how it works.

https://github.com/AveYo/Compressed2TXT

In fact, a process like 7-Zip or WinRAR is performed.

The choice is yours. You are free to use or not use this file.

In fact, if you examine the temp folder, you will see that it is the same as the previous files...

**Reply by abdullah-erturk** [reply]
Dear Zixim,

There is no confusing or suspicious code, all codes are open source and not powershell scripts. Runs as cmd script.

With the script, a powershell script is created and run in the temp folder, and finally it is deleted.

bat and cmd files are easier to run.

https://github.com/abdullah-erturk/RustDesk-ID-Changer

Code content:
https://github.com/abdullah-erturk/RustDesk-ID-Changer/blob/main/RustDesk_ID_Changer.bat

**Comment by Zixim** [comment]
For anyone interested, below is an all-in-one version of the script.
Copy/paste it into a file, save it with a .ps1 extension, run from an elevated (Administrator) Powershell.

<details>
  <summary>Here is an alternative script</summary>
  
  All-in-one plain text script.
```
function ChangeID($option) {
    Write-Host "`nChanging Rustdesk ID."

	# Stopping the service & closing the app
	Stop-Service RustDesk
	taskkill /im rustdesk.exe /f

	# Read the current id from Rustdesk configuration file
	$id = Get-Content "C:\Windows\ServiceProfiles\LocalService\AppData\Roaming\RustDesk\config\RustDesk.toml" | Select-Object -Index 0

	$newId = "id = '$option'"
	
	# Newer Rustdesk app has an encrypted ID, let's mention that to user
	if ( $id -like "*enc_id*") {$encrypted = "is encrypted"}
	
	# Show user what the ID is, and what it will become
	Write-Host "  Current ID $encrypted : $id"
	Write-Host "  New ID: $newId"
	
	# Read the Rustdesk configuration file
	$fileContent = Get-Content -Path "C:\Windows\ServiceProfiles\LocalService\AppData\Roaming\RustDesk\config\RustDesk.toml"

	# Replace the old ID with the new one
	$newContent = $fileContent -replace [regex]::Escape($id), $newId

	# Write the change into the Rustdesk configuration file
	$newContent | Set-Content -Path "C:\Windows\ServiceProfiles\LocalService\AppData\Roaming\RustDesk\config\RustDesk.toml"

	Restart-Service RustDesk
	
	Write-Host "DONE`n"	
}

# If we're not elevated, explain to user and exit.
if (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]::Administrator)) {
	Write-Output "This script needs to be run as Administrator because it needs to modify the Rustdesk configuration file."
	break
}

# Display menu options
Write-Host "Select an option:"
Write-Host "  1. Set RustDesk ID to the computer name:"$(get-content env:computername)
Write-Host "  2. Set RustDesk ID to a 9-digit random number"
Write-Host "  3. Set RustDesk ID to the value you specify"
Write-Host "  4. Exit"

while ($true) {
    # Prompt the user for input
    $choice = Read-Host "Enter the option of your choice (1-4)"

    switch ($choice) {
        '1' {
            ChangeID $(get-content env:computername)
        }
        '2' {
            ChangeID $(-join ((48..57) | Get-Random -Count 9 | ForEach-Object {[char]$_}))
        }
        '3' {
            ChangeID $(Read-Host "`n  Specify your custom RustDesk ID")
        }
        '4' {
            Write-Host "Exiting the script."
            exit
        }
        default {
            Write-Host "Invalid choice. Please select a valid option (1-4)."
        }
    }
	Exit
}
```
</details>

**Reply by TBG-FR** [reply]
You're a savior 👏
